### PR TITLE
Decode and verify UAA JWT tokens without connecting to UAA server

### LINKED ISSUE #3: Provide a way to validate tokens without connecting to an UAA server
Tokens are being checked with `/check_token` request to an UAA server.
This request can be replaced by decoding a token and JWS signature check.

### PR DESCRIPTION
**Pull request by hairyhum**
Fixes #3
Uses https://github.com/rabbitmq/uaa_jwt library to decode a token and verify a signature.
Signing keys should be predefined in the uaa_jwt application environment

### CODE REVIEW

**Comment by hairyhum** [comment]
Using https://github.com/rabbitmq/rabbitmq-common/pull/161 packages Elixir dependencies and Elixir itself into `*.ez` archives.

**Comment by michaelklishin** [comment]
This now has conflicts with master.

**Comment by hairyhum** [comment]
Rebased from `master`

File: src/rabbit_auth_backend_uaa.erl [hunk 1]
```diff
@@ -84,61 +88,25 @@ with_token(Token, Fun) ->
...
-    case Resp of
-        {ok, {{_HTTP, Code, _}, _Headers, Body}} ->
-            case Code of
-                200 -> parse_resp(Body);
-                400 -> parse_err(Body);
-                401 -> {error, resource_server_authentication_failed};
-                _   -> {error, {Code, Body}}
-            end;
-        {error, _} = E -> E
+    case 'Elixir.UaaJWT':decode_and_verify(Token) of
```
Comment (kjnilsson) [inline src/rabbit_auth_backend_uaa.erl:97]: Do we need to decode and verify the token each time? How about storing the decoded claims map in `impl` instead?

File: src/rabbit_auth_backend_uaa.erl [hunk 2]
```diff
@@ -23,7 +23,8 @@
 
 -export([description/0]).
 -export([user_login_authentication/2, user_login_authorization/1,
-         check_vhost_access/3, check_resource_access/3]).
+         check_vhost_access/3, check_resource_access/3,
+         check_topic_access/4]).
 
 %% httpc seems to get racy when using HTTP 1.1
 -define(HTTPC_OPTS, [{version, "HTTP/1.0"}]).
```
Comment (kjnilsson) [inline src/rabbit_auth_backend_uaa.erl]: is this used? likewise `description()` below refers to HTTP

**Review by kjnilsson — CHANGES_REQUESTED** [review]

File: src/rabbit_auth_backend_uaa.erl [hunk 3]
```diff
@@ -55,90 +53,61 @@ user_login_authorization(Username) ->
...
-                    rabbit_oauth2_scope:resource_access(Resource, Permission, Scopes)
-               end).
-
-check_topic_access(#auth_user{username = Username}, Resource, Permission, Context) ->
-    with_token(Username,
-               fun(UserData) ->
-                    Scopes = get_scopes(UserData),
-                    rabbit_oauth2_scope:topic_access(Resource, Permission, Context, Scopes)
-               end).
+check_vhost_access(#auth_user{username = Username, impl = DecodedToken},
```
Comment (kjnilsson) [inline src/rabbit_auth_backend_uaa.erl]: `Username` is no longer used.

**Review by kjnilsson — CHANGES_REQUESTED** [review]
`make tests` fail unless `make dist` is first run.

**Comment by hairyhum** [comment]
`test-dist` now is built before tests are run:
https://github.com/rabbitmq/rabbitmq-common/commit/60f5bb2579735cc74cddfd8cfe70fc980c85faa5

**Review by kjnilsson — APPROVED** [review]

**Pull request merged**